### PR TITLE
HBASE-22036 Rewrite TestScannerHeartbeatMessages

### PR DESCRIPTION
**Pull request by Apache9**
Introduced a `ScanPerNextResultScanner` to simulate the `ResultScanner` in the old time.

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ HBASE-21512 Compile Tests _ |
| +1 | mvninstall | 264 | HBASE-21512 passed |
| +1 | compile | 53 | HBASE-21512 passed |
| +1 | checkstyle | 70 | HBASE-21512 passed |
| +1 | shadedjars | 260 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 223 | HBASE-21512 passed |
| +1 | javadoc | 33 | HBASE-21512 passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 255 | the patch passed |
| +1 | compile | 55 | the patch passed |
| +1 | javac | 55 | the patch passed |
| +1 | checkstyle | 67 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 276 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 494 | Patch does not cause any errors with Hadoop 2.7.4 or 3.0.0. |
| +1 | findbugs | 225 | the patch passed |
| +1 | javadoc | 31 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 14001 | hbase-server in the patch failed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 16458 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/191 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 05c2201433f9 4.4.0-137-generic #163-Ubuntu SMP Mon Sep 24 13:14:43 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21512 / 14658419e4 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/1/testReport/ |
| Max. process+thread count | 5010 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 21 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ HBASE-21512 Compile Tests _ |
| +1 | mvninstall | 257 | HBASE-21512 passed |
| +1 | compile | 57 | HBASE-21512 passed |
| +1 | checkstyle | 75 | HBASE-21512 passed |
| +1 | shadedjars | 275 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 274 | HBASE-21512 passed |
| +1 | javadoc | 40 | HBASE-21512 passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 266 | the patch passed |
| -1 | compile | 32 | hbase-server in the patch failed. |
| -1 | javac | 32 | hbase-server in the patch failed. |
| +1 | checkstyle | 73 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 279 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 546 | Patch does not cause any errors with Hadoop 2.7.4 or 3.0.0. |
| +1 | findbugs | 257 | the patch passed |
| -1 | javadoc | 33 | hbase-server generated 4 new + 0 unchanged - 0 fixed = 4 total (was 0) |
||| _ Other Tests _ |
| -1 | unit | 36 | hbase-server in the patch failed. |
| +1 | asflicense | 11 | The patch does not generate ASF License warnings. |
| | | 2606 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/191 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux e43c2fdbab7c 4.4.0-131-generic #157~14.04.1-Ubuntu SMP Fri Jul 13 08:53:17 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21512 / dfa4f47b59 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| compile | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/2/artifact/out/patch-compile-hbase-server.txt |
| javac | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/2/artifact/out/patch-compile-hbase-server.txt |
| javadoc | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/2/artifact/out/diff-javadoc-javadoc-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/2/testReport/ |
| Max. process+thread count | 86 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 290 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ HBASE-21512 Compile Tests _ |
| +1 | mvninstall | 327 | HBASE-21512 passed |
| +1 | compile | 68 | HBASE-21512 passed |
| +1 | checkstyle | 80 | HBASE-21512 passed |
| +1 | shadedjars | 347 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 296 | HBASE-21512 passed |
| +1 | javadoc | 41 | HBASE-21512 passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 306 | the patch passed |
| +1 | compile | 69 | the patch passed |
| +1 | javac | 69 | the patch passed |
| +1 | checkstyle | 85 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 332 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 685 | Patch does not cause any errors with Hadoop 2.7.4 or 3.0.0. |
| +1 | findbugs | 320 | the patch passed |
| +1 | javadoc | 44 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 14937 | hbase-server in the patch passed. |
| +1 | asflicense | 32 | The patch does not generate ASF License warnings. |
| | | 18338 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/191 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 3b6e3628cced 4.4.0-145-generic #171-Ubuntu SMP Tue Mar 26 12:43:40 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21512 / dfa4f47b59 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/3/testReport/ |
| Max. process+thread count | 4761 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-191/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/client/ScanPerNextResultScanner.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import org.apache.hbase.thirdparty.com.google.common.base.Throwables;
+
+/**
+ * A ResultScanner which will only send request to RS when there are no cached results when calling
+ * next, just like the ResultScanner in the old time. Mainly used for writing UTs, that we can
+ * control when to send request to RS. The default ResultScanner implementation will fetch in
+ * background.
+ */
+@InterfaceAudience.Private
+public class ScanPerNextResultScanner implements ResultScanner, AdvancedScanResultConsumer {
```
Comment (openinx) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/client/ScanPerNextResultScanner.java]: Seems this class is only used for testing ?  I prefer to move this into test package, or other future packages will use this ?

Comment (Apache9) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/client/ScanPerNextResultScanner.java]: It is under the src/test/java?

Comment (openinx) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/client/ScanPerNextResultScanner.java]: OK, my bad..

**Pull request merged**